pkg/controller/datavolume/external-population-controller.go

```diff
@@ -0,0 +1,206 @@
+}
+
+// Reconcile loop for externally populated DataVolumes
+// Unlike the base class sync call, we use a prepare function to populate the otherwise nil source field
+func (r PopulatorReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("DataVolume", req.NamespacedName)
+	res, syncErr := r.sync(log, req, nil, r.prepare)
+	return r.updateStatus(log, res, syncErr)
+}
+
+// If dataSource/dataSourceRef is set (external population), use an empty spec.Source field
+func (r PopulatorReconciler) prepare(dv *cdiv1.DataVolume) error {
+	if !dvUsesVolumePopulator(dv) {
+		return errors.Errorf("undefined population source")
+	}
+	dv.Spec.Source = &cdiv1.DataVolumeSource{}
+	return nil
+}
+
+// checkPopulationRequirements returns true if the PVC meets the requirements to be populated
+func (r *PopulatorReconciler) checkPopulationRequirements(pvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume, event *Event) (bool, error) {
+	csiDriverAvailable, err := r.storageClassCSIDriverExists(pvc.Spec.StorageClassName)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return false, err
+	}
+	// Non-snapshot population will only work with CSI drivers
+	if !isSnapshotPopulation(pvc) && !csiDriverAvailable {
+		r.recorder.Event(pvc, corev1.EventTypeWarning, NoCSIDriverForExternalPopulation, MessageNoCSIDriverForExternalPopulation)
+		event.reason = NoCSIDriverForExternalPopulation
+		event.message = MessageNoCSIDriverForExternalPopulation
+		event.eventType = corev1.EventTypeWarning
+		return false, nil
+	}
+
+	return r.checkAnyVolumeDataSource(pvc, dv, event)
+}
+
+// checkAnyVolumeDataSource determines if the AnyVolumeDataSource feature gate is enabled or not
+func (r *PopulatorReconciler) checkAnyVolumeDataSource(pvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume, event *Event) (bool, error) {
+	// We don't need the AnyVolumeDataSource feature gate if the PVC
+	// is externally populated by a PVC or Snapshot
+	nonPopulator := isPvcPopulation(pvc) || isSnapshotPopulation(pvc)
+
+	// If the AnyVolumeDataSource feature gate is disabled, Kubernetes drops the contents of the dataSourceRef field.
+	// We can then determine if the feature is enabled or not by checking that field after creating the PVC.
+	enabled := pvc.Spec.DataSourceRef != nil
+	if !enabled && !nonPopulator {
+		r.recorder.Event(pvc, corev1.EventTypeWarning, NoAnyVolumeDataSource, MessageNoAnyVolumeDataSource)
+		event.reason = NoAnyVolumeDataSource
+		event.message = MessageNoAnyVolumeDataSource
+		event.eventType = corev1.EventTypeWarning
+	} else {
+		event.reason = ExternalPopulationSucceeded
+		event.message = fmt.Sprintf(MessageExternalPopulationSucceeded, pvc.Name, pvc.Spec.DataSourceRef.Name)
+		event.eventType = corev1.EventTypeNormal
+	}
+
+	return enabled, nil
+}
+
+// dvUsesVolumePopulator returns true if the datavolume's PVC is meant to be externally populated
+func dvUsesVolumePopulator(dv *cdiv1.DataVolume) bool {
```

Comment:
for consistency: is(Dv)UsingVolumePopulator, isPopulatedDataSource/Ref etc.

Reply:
For `isPopulatedDataSource/Ref` I agree it may be better the proposed change, but I actually think `dvUsesVolumePopulator` is more accurate.